lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp

```diff
@@ -601,40 +601,75 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
                   return success();
                 });
+  // onnx.Clip with min/max as arguments introduced in opset 11
   patterns.onOp(
-      "Clip", 13, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+      "Clip", 11, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        // https://onnx.ai/onnx/operators/onnx__Clip.html
+        // Note: version 13 added 'bfloat16' support
+        // Note: version 12 added 'int16', 'int32', etc. support
+        // We could check each of those against the specific versions... or
+        // just let them through.
+
+        // Inputs and outputs must be tensors. Min and max are both optional
+        // and if provided must be "tensors of empty shape" (scalars).
+        Value source;
         Torch::ValueTensorType resultType;
+        if (binder.tensorOperandAtIndex(source, 0) ||
+            binder.tensorResultType(resultType)) {
+          return failure();
+        }
+
         if (binder.op->getNumOperands() == 1) {
-          Value source;
-          if (binder.tensorOperand(source) ||
-              binder.tensorResultType(resultType))
-            return failure();
-          Value cstNone =
-              rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
-          rewriter.replaceOpWithNewOp<Torch::AtenClampOp>(
-              binder.op, resultType, source, /*min=*/cstNone, /*max=*/cstNone);
+          // Cliping between numeric_limits::lowest() and numeric_limits::max()
+          // is a no-op.
+          rewriter.replaceOp(binder.op, source);
           return success();
         } else if (binder.op->getNumOperands() == 2) {
-          Value source, min;
-          if (binder.tensorOperands(source, min) ||
-              binder.tensorResultType(resultType))
-            return failure();
           rewriter.replaceOpWithNewOp<Torch::AtenClampMinTensorOp>(
-              binder.op, resultType, source, /*min=*/min);
+              binder.op, resultType, source, /*min=*/binder.op->getOperand(1));
           return success();
         } else if (binder.op->getNumOperands() == 3) {
-          Value source, min, max;
-          if (binder.tensorOperandAtIndex(source, 0) ||
-              binder.tensorOperandAtIndex(min, 1) ||
-              binder.tensorOperandAtIndex(max, 2) ||
-              binder.tensorResultType(resultType))
-            return failure();
           rewriter.replaceOpWithNewOp<Torch::AtenClampTensorOp>(
-              binder.op, resultType, source, min, max);
+              binder.op, resultType, source, /*min=*/binder.op->getOperand(1),
+              /*max=*/binder.op->getOperand(2));
           return success();
         }
         return failure();
       });
+  // onnx.Clip with min/max as attributes
+  patterns.onOp(
+      "Clip", 1, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        // https://onnx.ai/onnx/operators/onnx__Clip.html
+        // Note: version 6 has optional attributes with default min/max values.
```

Comment:
(I'd still prefer to drop this - opset 11 is old, and it doesn't seem worth duplicating logic already in the version converter)

Reply:
You don't have to duplicate an additional rewriter, you just insert a check for the attributes and operand creator. Generally we want to avoid recursive invocation of rewriters when possible.

https://github.com/llvm/torch-mlir/blob/main/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp#L1178

I can promise you we need to support this. Some of the models I have debugged have already had the attribute based passing mechanism. Relying on the external tooling to upgrade will not guarantee support. The `ReduceMin` example used attributes for defining the axes despite being version 13. Old != unused.

Reply:
> You don't have to duplicate an additional rewriter, you just insert a check for the attributes and operand creator.

Alright, I can try to clean up https://github.com/ScottTodd/torch-mlir/commit/b6c307b023b52adc3a46e09c33c16298fc6d52e9 ... it's not pretty :/